File: plugins/harness/skills/harness-install/assets/uv/runtime/rules/ast/import_over_fqn.py
#!/usr/bin/env -S uv run
# -*- coding: utf-8 -*-
"""
Require import over FQN rule.
"""

import sys


from collections.abc import Iterable

import libcst as cst

from rules.harness_check_rule import (
    Finding,
    FindingFix,
    FixSafety,
    HarnessCheckRule,
    parse_python,
)
from core.rule_context import RuleContext, relative, stack_sources_configured

sys.stdout.reconfigure(encoding="utf-8")
sys.stderr.reconfigure(encoding="utf-8")


class ImportOverFqnRule(HarnessCheckRule):
    """Validate importOverFqn check."""

    category = "importOverFqn"

    def applies(self, ctx: RuleContext) -> bool:
        """Check if this rule applies to the context."""
        if not ctx.is_enabled(self.category):
            return False
        return stack_sources_configured(ctx.manifest.raw, self.category)

    def validate(self, ctx: RuleContext) -> Iterable[Finding]:
        """Validate importOverFqn check."""
        severity = ctx.severity_of(self.category)
        category = self.category
        sources = ctx.stack_sources(self.category)

        class FqnFinder(cst.CSTVisitor):
            METADATA_DEPENDENCIES = (cst.metadata.PositionProvider,)

            def __init__(self, rel_path: str) -> None:
                super().__init__()
                self.findings: list[Finding] = []
                self.rel_path = rel_path
                self.imported_names = set()

            def visit_ImportFrom(self, node: cst.ImportFrom) -> bool:
                if not isinstance(node.names, cst.ImportStar):
                    names_seq = (
                        node.names
                        if isinstance(node.names, (list, tuple))
                        else [node.names]
                    )
                    for name_item in names_seq:
                        if isinstance(name_item, cst.ImportAlias):
                            self.imported_names.add(
                                name_item.name.value
                                if isinstance(name_item.name, cst.Name)
                                else str(name_item.name)
                            )
                return True

            def visit_Attribute(self, node: cst.Attribute) -> bool:
                depth = 0
                current = node
                while isinstance(current, cst.Attribute):
                    depth += 1
                    current = current.value
                if depth >= 2 and isinstance(current, cst.Name):
                    fqn_parts = [current.value]
                    current = node
                    while isinstance(current, cst.Attribute):
                        fqn_parts.append(current.attr.value)
                        current = current.value
                    fqn_parts.reverse()
                    simple_name = fqn_parts[0]
                    if simple_name not in self.imported_names:
                        fqn_str = ".".join(fqn_parts)
                        pos = self.get_metadata(cst.metadata.PositionProvider, node)
                        self.findings.append(
                            Finding(
                                severity,
                                category,
                                f"{self.rel_path}:{pos.start.line}: fully qualified name `{fqn_str}` used inline; add an import and use the simple name",
                                file=self.rel_path,
                                start_line=pos.start.line,
                                start_column=pos.start.column + 1,
                                end_line=pos.end.line,
                                end_column=pos.end.column + 1,
                                fix=FindingFix(
                                    description="add import statement for top-level module; may alter resolution",
                                    safety=FixSafety.UNSAFE,
                                    edits=(),
                                ),
                            )
                        )
                return True

        def collect_findings():
            for path in sources:
                tree, error = parse_python(path)
                if error is not None:
                    yield Finding(
                        severity,
                        category,
                        f"{relative(path, ctx.root)}: syntax error: {error}",
                    )
                    continue
                wrapper = cst.MetadataWrapper(tree)
                visitor = FqnFinder(relative(path, ctx.root))
                wrapper.visit(visitor)
                yield from visitor.findings
        return list(collect_findings())


RULE: HarnessCheckRule = ImportOverFqnRule()

File: plugins/harness/skills/harness-install/assets/bun/runtime/rules/ast/import-over-fqn.ts
#!/usr/bin/env bun
// -*- coding: utf-8 -*-
import type { Node, SourceFile } from "typescript@6.0.3";
import {
    createSourceFile,
    isIdentifier,
    isImportDeclaration,
    isPropertyAccessExpression,
    SyntaxKind,
} from "typescript@6.0.3";
import { astChildrenOf } from "../../core/ast-traversal";
import type { Finding, HarnessCheckRule, RuleContext } from "../harness-check-rule";

/**
 * Require imports instead of fully qualified names. When the simple name from an
 * import is available, use it instead of inline FQN. If simple name conflicts with
 * another import, FQN is allowed.
 */
export const importOverFqnRule: HarnessCheckRule = {
    category: "importOverFqn",
    applies(ctx: RuleContext): boolean {
        const section = ctx.manifest.raw.importOverFqn;
        if (typeof section !== "object" || section === null) {
            return false;
        }
        const enabled = (section as { enabled?: unknown }).enabled;
        return enabled !== false;
    },
    validate(ctx: RuleContext): readonly Finding[] {
        return ctx.stackSources("importOverFqn").flatMap((file) => {
            const text = ctx.read(file);
            if (!text) {
                return [];
            }
            const sourceFile: SourceFile = createSourceFile(file, text, SyntaxKind.LatestVersion, true);
            const importedNames = new Set<string>();
            const collectImports = (node: Node): void => {
                if (
                    isImportDeclaration(node) &&
                    node.importClause?.namedBindings &&
                    !("name" in node.importClause.namedBindings)
                ) {
                    const bindings = node.importClause.namedBindings;
                    if ("elements" in bindings) {
                        for (const element of bindings.elements) {
                            importedNames.add(element.name.text);
                        }
                    }
                }
                astChildrenOf(node).forEach(collectImports);
            };
            collectImports(sourceFile);
            const checkFqn = (node: Node): readonly Finding[] => {
                if (isPropertyAccessExpression(node) && node.expression) {
                    const traverseDepth = (curr: Node | undefined, d: number): [number, Node | undefined] => {
                        if (!curr || !isPropertyAccessExpression(curr)) {
                            return [d, curr];
                        }
                        return traverseDepth(curr.expression, d + 1);
                    };
                    const [depth, current] = traverseDepth(node, 0);
                    if (depth >= 2 && current && isIdentifier(current)) {
                        const simpleName = current.text;
                        if (!importedNames.has(simpleName)) {
                            const buildFqnParts = (curr: Node | undefined, acc: string[]): string[] => {
                                if (!curr || !isPropertyAccessExpression(curr)) {
                                    return acc.reverse();
                                }
                                return buildFqnParts(curr.expression, acc.concat(curr.name.text));
                            };
                            const start = sourceFile.getLineAndCharacterOfPosition(node.getStart(sourceFile));
                            const end = sourceFile.getLineAndCharacterOfPosition(node.getEnd());
                            return [
                                {
                                    severity: ctx.severityOf("importOverFqn"),
                                    category: "importOverFqn",
                                    message: `fully qualified name \`${buildFqnParts(node, [simpleName]).join(".")}\` used inline; add an import and use the simple name`,
                                    file,
                                    startLine: start.line + 1,
                                    startColumn: start.character + 1,
                                    endLine: end.line + 1,
                                    endColumn: end.character + 1,
                                    fix: {
                                        description: `add import for \`${simpleName}\` and replace FQN with simple name`,
                                        safety: "unsafe",
                                        edits: [],
                                    },
                                },
                            ].concat(astChildrenOf(node).flatMap(checkFqn));
                        }
                    }
                }
                return astChildrenOf(node).flatMap(checkFqn);
            };
            return checkFqn(sourceFile);
        });
    },
};
